doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,47 @@
+# Design Doc: How to Execute the ProgramDesc on Multiple Device
+
+A sequence of optimizers will convert a ProgramDesc to ExecutionPlan, which allow
+the PaddlePaddle program running on multi devices such CPU, GPU or FPGA,
+or multi nodes.
+
+The interface of ExecutionPlan is as followed:
+
+```cpp
+enum ExecutionType {
+    SINGLE_THREAD = 0,
+    MULTIPLE_CPU = 1,
+    MULTIPLE_GPU = 2,
+    MULTIPLE_NODE = 3
+};
+
+class ExecutionPlan {
+  ExecutionType type_;
+  vector<ProgramDescBind*> trainers_;
+  vector<ProgramDescBind*> pservers_;
+};
+```
+
+## Running with Multi-CPU
+
+<img src="src/multi_cpu.png">
+
+1. Use Multi-CPU Optimizer to convert the ProgramDesc to ExecutionPlan
+
+    For the data parallelism, we need to pass the attribution `start` and `end`
+    index for the mini-batch, and this will be calculated in the optimizer step.
+1. `multiCPUExecutor` execute the ExecutionPlan which the type is Multi CPU
```

Comment:
We don't need a `multiCPUExecutor`, just check `trainer_count` in the current `Executor` and do the copy if `trainer_count > 1`

Reply:
same with @typhoonzero . I think that we do not need a `MultiCPUExecutor`, since every thread can visit the same memory, I can not see any benefit to split thread execution into separated scope/block.
And how should it be when the math library has built-in openmp.  Should we classified it as `single thread executor` or `MultiCPUExecutor`?

Reply:
As the comment: https://github.com/PaddlePaddle/Paddle/pull/6223#discussion_r154692877, for my understand, each thread will calculate the gradient independently, and the graident has the same variable name, so maybe they should be create on different scope, or maybe my understanding has something wrong?

Reply:
@Yancey1989 you are right, scopes can have child scopes to resolve this.

Reply:
Definitely, at least, there should be **different gradients variables**(not only different name). Scope is just grouping them in somewhere -- for example, created different gradients variables with specific name prefix in global scope is a visible solution. Create a new `Scope` or not isn't the key issue to this concept. 

I believe that our `Executor` should have the ability to run in multithread enviroment. We do not need the concept of `MultiCPUExecutor`.

Reply:
@dzhwinter 
> And how should it be when the math library has built-in openmp

I think it's independent with the data parallelism, users could set the thread count of openmp or the threads of executor by themselves.

> Create a new Scope or not isn't the key issue to this concept

Got it, and if we need the multiple variable with different prefix,  The idea of @helinwang is nice, https://github.com/PaddlePaddle/Paddle/pull/6223#issuecomment-349088250

Reply:
@dzhwinter 
> I believe that our Executor should have the ability to run in multithread enviroment. We do not need the concept of MultiCPUExecutor.

It's a good idea, we can pass a parameter `thread_count` to `Executor`, and the `Executor` will execute the ExecutionPlan with multi threads. For another, I think we also need to insert some sync Op while `Split X` and `Reduce dW`.